EasyCommands/CommandParsers/ParameterProcessors.cs

```diff
@@ -116,1146 +319,307 @@ public static class ParameterProcessorRegistry {
+        static DataProcessor<T> rightList<T>(bool required) where T: class, CommandParameter {
+            return new DataProcessor<T>(new ListValueDataFetcher<T>(required), df => p => false, df => p => df.SetValue(p));
         }
 
-        public class VariableSelectorProcessor : SimpleParameterProcessor<VariableSelectorCommandParameter> {
-            BlockType? blockType = null;
-            bool isGroup = false;
-
-            public override bool CanConvert(List<CommandParameter> p) {
-                return blockType != null;
-            }
-
-            public override CommandParameter Convert(List<CommandParameter> p) {
-                VariableSelectorCommandParameter selector = findFirst<VariableSelectorCommandParameter>(p);
-                Variable variable = selector.Value;
-                return new SelectorCommandParameter(new SelectorEntityProvider(blockType.Value, isGroup, variable));
-            }
-
-            public override void Initialize() {
-                blockType = null;
-                isGroup = false;
-            }
-
-            public override bool ProcessLeft(CommandParameter p) {
-                return false;
-            }
-
-            public override bool ProcessRight(CommandParameter p) {
-                if (p is BlockTypeCommandParameter && blockType == null) blockType = ((BlockTypeCommandParameter)p).Value;
-                else if (p is GroupCommandParameter && !isGroup) isGroup = true;
-                else return false;
-                return true;
-            }
+        //ParameterProcessors
+        public interface ParameterProcessor {
+            List<Type> GetProcessedTypes();
+            bool CanProcess(CommandParameter p); 
+            bool Process(List<CommandParameter> p, int i, out List<CommandParameter> finalParameters);
+            int GetRank();
+            void SetRank(int rank);
         }
 
-        public class PrintCommandProcessor : SimpleParameterProcessor<PrintCommandParameter> {
-            Variable variable;
-
-            public override bool CanConvert(List<CommandParameter> p) {
-                return variable != null;
-            }
-
-            public override CommandParameter Convert(List<CommandParameter> p) {
-                return new CommandReferenceParameter(new PrintCommand(variable));
-            }
-
-            public override void Initialize() {
-                variable = null;
+        public abstract class ParameterProcessor<T> : ParameterProcessor where T : class, CommandParameter {
+            int rank;
+            public virtual List<Type> GetProcessedTypes() { return new List<Type>() { typeof(T) }; }
+            public virtual int GetRank() => rank;
+            public virtual void SetRank(int r) => rank = r;
+            public bool CanProcess(CommandParameter p) {
+                return p is T;
             }
+            public abstract bool Process(List<CommandParameter> p, int i, out List<CommandParameter> finalParameters);
```

Comment:
With rank living in `ParameterProcessor`, you could implement `IComparable<ParameterProcessor>` and have:

```c#
public int CompareTo(ParameterProcessor other)
{
    return Rank.compareTo(other.Rank);
}
```

Then, you can delete your custom insertion logic up above and just do `parameterProcessors.sort()` to keep it sorted. (The `sort` method, by default, looks to see if the objects in the list implement `IComparable` or `IComparer`, and if so, uses its given comparison method to perform the sorting)

Reply:
Great call out, I've made ParameterProcessor implement IComparable.

I'm struggling on how to add attributes to the ParameterProcessor interface.  Making the interface an abstract class adds more characters (need to specify "public", "abstract", "override" keywords which can't be minified)  

Maybe you've got some ideas on how to simplify?

For now I condensed as much as I could but get & set methods are still present on the interface. :(